File: ts/packages/core/src/mapper.ts
// The canonical ilmek-event → mekik-frame mapping (PROTOCOL.md §4).
//
// This is the closed, golden-tested core of the protocol. `TurnMapper` consumes
// one run's `IlmekEvent` stream and produces the mekik frames a client sees. It
// is turn-stateful - it owns the current turn's genui stream id and chunk
// counter - but every source of nondeterminism (the persistent-seq allocator,
// the id minter, the clock, the reply selector) is injected, so the fixtures in
// conformance/mekik/fixtures replay byte-for-byte across TS and .NET.

import { isToken, type IlmekEvent, type Pending } from "@ilmek/core";

import type {
    AIChunk,
    GenUIFrame,
    InterruptFrame,
    MessageAction,
    OutgoingFrame,
    TextOutFrame,
    ToolCall,
    ToolCallFrame,
    UiRef,
} from "./protocol.ts";

/** Mints the ids that appear on the wire. Production: random. Fixtures: 1-based counters. */
export interface IdMinter {
    /** For `text` frame `id`. */
    message(): string;
    /** For a turn's genui `streamId`. */
    stream(): string;
}

export interface TurnMapperDeps {
    /** Advance and return the conversation's next persistent `seq`. */
    allocSeq: () => number;
    mint: IdMinter;
    /** Wall clock (ms). Injected so fixtures pin `timestamp`. */
    now: () => number;
    /**
     * Pick the run's reply text from final channel state at `run_end{done}`
     * (PROTOCOL.md §4.3). Returns undefined/"" to emit no consolidated text.
     */
    reply?: (state: Record<string, unknown>) => string | undefined;
}

/** The reserved key `mekik.approve`/`mekik.ui`/`mekik.tool` tuck metadata under. */
const MEKIK_KEY = "$mekik";

interface MekikGenUIPayload {
    [MEKIK_KEY]: "genui";
    chunk: AIChunk;
}
interface MekikToolPayload {
    [MEKIK_KEY]: "tool";
    call: ToolCall;
}

function isRecord(v: unknown): v is Record<string, unknown> {
    return typeof v === "object" && v !== null && !Array.isArray(v);
}
function isGenUIPayload(v: unknown): v is MekikGenUIPayload {
    return isRecord(v) && v[MEKIK_KEY] === "genui" && isRecord(v.chunk);
}
function isToolPayload(v: unknown): v is MekikToolPayload {
    return isRecord(v) && v[MEKIK_KEY] === "tool" && isRecord(v.call);
}

/** Split an interrupt payload into its client-facing parts (PROTOCOL.md §4.2). */
export function unwrapInterrupt(payload: unknown): { payload: unknown; ui?: UiRef; actions?: MessageAction[] } {
    if (!isRecord(payload) || !isRecord(payload[MEKIK_KEY])) {
        return { payload };
    }
    const meta = payload[MEKIK_KEY] as { ui?: UiRef; actions?: MessageAction[] };
    // Strip the reserved key from what the human sees; keep everything else.
    const rest: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(payload)) if (k !== MEKIK_KEY) rest[k] = v;
    return {
        payload: rest,
        ...(meta.ui !== undefined ? { ui: meta.ui } : {}),
        ...(meta.actions !== undefined ? { actions: meta.actions } : {}),
    };
}

/** Build the `interrupt` frame (or its `welcome.pending` view minus seq) for one pending pause. */
export function interruptFrameData(p: Pending): InterruptFrame["data"] {
    const { payload, ui, actions } = unwrapInterrupt(p.payload);
    return {
        payload,
        ...(ui !== undefined ? { ui } : {}),
        ...(actions !== undefined ? { actions } : {}),
    };
}

export class TurnMapper {
    private readonly deps: TurnMapperDeps;
    private streamId: string | null = null;
    private chunkCounter = 0;

    constructor(deps: TurnMapperDeps) {
        this.deps = deps;
    }

    /** Frames for one ilmek event, in emit order. May be empty (dropped events). */
    map(ev: IlmekEvent): OutgoingFrame[] {
        switch (ev.type) {
            case "run_start":
                return [{ type: "run", data: { status: "started" } }];

            case "custom":
                return this.mapCustom(ev.payload);

            case "interrupt":
                return ev.pending.map((p) => this.interruptFrame(p));

            case "run_end":
                return this.mapRunEnd(ev);

            // Not surfaced in v1 (PROTOCOL.md §4.1). A future `debug` mode may map these.
            case "node_start":
            case "node_end":
            case "node_error":
            case "node_retry":
            case "step_start":
            case "state":
            case "checkpoint":
                return [];
        }
    }

    private mapCustom(payload: unknown): OutgoingFrame[] {
        if (isToken(payload)) {
            return [this.genuiFrame({ type: "text", content: payload.text }, false)];
        }
        if (isGenUIPayload(payload)) {
            return [this.genuiFrame(payload.chunk, false)];
        }
        if (isToolPayload(payload)) {
            const frame: ToolCallFrame = { type: "tool_call", seq: this.deps.allocSeq(), data: payload.call };
            return [frame];
        }
        // Unrecognised customs are dropped here; an extension hook (MekikOptions
        // .onCustom) maps them in the engine, outside this closed core.
        return [];
    }

    private mapRunEnd(ev: Extract<IlmekEvent, { type: "run_end" }>): OutgoingFrame[] {
        switch (ev.status) {
            case "interrupted":
                // The interrupt frames were already emitted from the `interrupt`
                // event; here we only close the run.
                return [{ type: "run", data: { status: "interrupted" } }];

            case "aborted":
                // No text: the last checkpoint stands and the thread is resumable.
                return [{ type: "run", data: { status: "aborted" } }];

            case "error": {
                const message = formatErrors(ev.errors);
                const text = this.botText(`⚠️ ${message}`);
                return [text, { type: "run", data: { status: "error" } }];
            }

            case "done": {
                const out: OutgoingFrame[] = [];
                if (this.streamId !== null) {
                    out.push(this.genuiFrame({ type: "event", name: "stream_done" }, true));
                }
                const reply = this.deps.reply?.(ev.state);
                if (typeof reply === "string" && reply.length > 0) out.push(this.botText(reply));
                out.push({ type: "run", data: { status: "finished" } });
                return out;
            }
        }
    }

    private interruptFrame(p: Pending): InterruptFrame {
        return { type: "interrupt", seq: this.deps.allocSeq(), id: p.id, data: interruptFrameData(p) };
    }

    private genuiFrame(chunk: AIChunk, done: boolean): GenUIFrame {
        if (this.streamId === null) this.streamId = this.deps.mint.stream();
        // Assign a chunk id if the emitter didn't; ids increment within the stream.
        const withId: AIChunk = chunk.id === undefined ? { ...chunk, id: this.nextChunkId() } : chunk;
        return { type: "genui", seq: this.deps.allocSeq(), streamId: this.streamId, done, chunk: withId };
    }

    private botText(text: string): TextOutFrame {
        return {
            type: "text",
            id: this.deps.mint.message(),
            seq: this.deps.allocSeq(),
            from: "bot",
            data: { text },
            timestamp: this.deps.now(),
        };
    }

    private nextChunkId(): number {
        return ++this.chunkCounter;
    }
}

function formatErrors(errors: ReadonlyArray<readonly [string, unknown]>): string {
    if (errors.length === 0) return "the run failed";
    return errors
        .map(([node, err]) => {
            const msg = err instanceof Error ? err.message : String(err);
            return `${node}: ${msg}`;
        })
        .join("; ");
}

/**
 * Drive a whole recorded event list through a fresh mapper (the fixture entry
 * point). Production streams events one at a time through `TurnMapper.map`; this
 * is the batch form the golden runner and `run()`-style callers use.
 */
export function eventToFrames(events: readonly IlmekEvent[], deps: TurnMapperDeps): OutgoingFrame[] {
    const mapper = new TurnMapper(deps);
    const out: OutgoingFrame[] = [];
    for (const ev of events) out.push(...mapper.map(ev));
    return out;
}

File: dotnet/src/Mekik.Core/Mapper.cs
using Ilmek;

namespace Mekik;

/// <summary>Mints the ids that appear on the wire. Production: random. Fixtures: 1-based counters.</summary>
public interface IIdMinter
{
    /// <summary>For a `text` frame `id`.</summary>
    string Message();
    /// <summary>For a turn's genui `streamId`.</summary>
    string Stream();
}

/// <summary>The injected environment a <see cref="TurnMapper"/> needs (PROTOCOL.md §4).</summary>
public sealed record TurnMapperDeps
{
    /// <summary>Advance and return the conversation's next persistent `seq`.</summary>
    public required Func<long> AllocSeq { get; init; }
    public required IIdMinter Mint { get; init; }
    /// <summary>Wall clock (ms). Injected so fixtures pin `timestamp`.</summary>
    public required Func<long> Now { get; init; }
    /// <summary>Pick the run's reply text from final channel state at `run_end{done}` (PROTOCOL.md §4.3).</summary>
    public Func<IReadOnlyDictionary<string, object?>, string?>? Reply { get; init; }
}

/// <summary>
/// The canonical ilmek-event → mekik-frame mapping (PROTOCOL.md §4), mirror of
/// the TypeScript <c>TurnMapper</c>. Turn-stateful (owns the genui stream id and
/// chunk counter); every nondeterministic input is injected so the golden
/// fixtures replay identically across the two languages. Frames are built as
/// <c>Dictionary&lt;string, object?&gt;</c> for byte-exact wire parity.
/// </summary>
public sealed class TurnMapper
{
    private const string MekikKey = "$mekik";

    private readonly TurnMapperDeps _deps;
    private string? _streamId;
    private long _chunkCounter;
    // The id of the open text run, or null when none is open. Consecutive text
    // deltas share it so the client renders one growing bubble, not one per token.
    private long? _textRunId;

    public TurnMapper(TurnMapperDeps deps) => _deps = deps;

    /// <summary>Frames for one ilmek event, in emit order. May be empty (dropped events).</summary>
    public IReadOnlyList<Dictionary<string, object?>> Map(IlmekEvent ev) => ev switch
    {
        RunStartEvent => [Run("started")],
        CustomEvent c => MapCustom(c.Payload),
        InterruptEvent i => i.Pending.Select(InterruptFrame).ToList(),
        RunEndEvent r => MapRunEnd(r),
        // Not surfaced in v1 (PROTOCOL.md §4.1): node/step/state/checkpoint events.
        _ => [],
    };

    private List<Dictionary<string, object?>> MapCustom(object? payload)
    {
        if (TryTokenText(payload, out var text))
            return [GenuiFrame(new Dictionary<string, object?> { ["type"] = "text", ["content"] = text }, false)];

        if (payload is IReadOnlyDictionary<string, object?> d)
        {
            if (d.GetValueOrDefault(MekikKey) is "genui" && d.GetValueOrDefault("chunk") is IReadOnlyDictionary<string, object?> chunk)
                return [GenuiFrame(chunk, false)];

            if (d.GetValueOrDefault(MekikKey) is "tool" && d.GetValueOrDefault("call") is IReadOnlyDictionary<string, object?> call)
                return [new Dictionary<string, object?> { ["type"] = "tool_call", ["seq"] = _deps.AllocSeq(), ["data"] = call }];
        }

        // Unrecognised customs are dropped here (an extension hook maps them outside this closed core).
        return [];
    }

    private List<Dictionary<string, object?>> MapRunEnd(RunEndEvent ev)
    {
        switch (ev.Status)
        {
            case RunStatus.Interrupted:
                // The interrupt frames were already emitted from the `interrupt` event.
                return [Run("interrupted")];
            case RunStatus.Aborted:
                // No text: the last checkpoint stands and the thread is resumable.
                return [Run("aborted")];
            case RunStatus.Error:
                return [BotText("⚠️ " + FormatErrors(ev.Errors)), Run("error")];
            case RunStatus.Done:
            default:
                var frames = new List<Dictionary<string, object?>>();
                if (_streamId is not null)
                    frames.Add(GenuiFrame(new Dictionary<string, object?> { ["type"] = "event", ["name"] = "stream_done" }, true));
                var reply = _deps.Reply?.Invoke(ev.FinalState ?? EmptyState);
                if (!string.IsNullOrEmpty(reply)) frames.Add(BotText(reply));
                frames.Add(Run("finished"));
                return frames;
        }
    }

    private Dictionary<string, object?> InterruptFrame(Pending p) => new()
    {
        ["type"] = "interrupt",
        ["seq"] = _deps.AllocSeq(),
        ["id"] = p.Id,
        ["data"] = InterruptFrameData(p),
    };

    private Dictionary<string, object?> GenuiFrame(IReadOnlyDictionary<string, object?> chunk, bool done)
    {
        _streamId ??= _deps.Mint.Stream();
        // Clone, then give the chunk its stream-scoped id. Consecutive text deltas
        // share ONE id so a client renders one growing bubble instead of one bubble
        // per token (PROTOCOL.md §4.1); a ui/event chunk — or a caller-supplied id —
        // closes the open text run, so the next text delta starts a fresh bubble.
        var withId = chunk.ToDictionary(kv => kv.Key, kv => kv.Value);
        if (withId.GetValueOrDefault("id") is not null)
        {
            _textRunId = null; // an explicit id opts out of run coalescing
        }
        else if (withId.GetValueOrDefault("type") as string == "text")
        {
            _textRunId ??= ++_chunkCounter;
            withId["id"] = _textRunId.Value;
        }
        else
        {
            _textRunId = null; // a non-text chunk ends the current text run
            withId["id"] = ++_chunkCounter;
        }
        return new Dictionary<string, object?>
        {
            ["type"] = "genui",
            ["seq"] = _deps.AllocSeq(),
            ["streamId"] = _streamId,
            ["done"] = done,
            ["chunk"] = withId,
        };
    }

    private Dictionary<string, object?> BotText(string text) => new()
    {
        ["type"] = "text",
        ["id"] = _deps.Mint.Message(),
        ["seq"] = _deps.AllocSeq(),
        ["from"] = "bot",
        ["data"] = new Dictionary<string, object?> { ["text"] = text },
        ["timestamp"] = _deps.Now(),
    };

    private static Dictionary<string, object?> Run(string status) => new()
    {
        ["type"] = "run",
        ["data"] = new Dictionary<string, object?> { ["status"] = status },
    };

    // ── interrupt payload wrapping (PROTOCOL.md §4.2) ─────────────────────────

    /// <summary>Split an interrupt payload into `{ payload, ui?, actions? }` (public for `welcome.pending`).</summary>
    public static Dictionary<string, object?> InterruptFrameData(Pending p) => Unwrap(p.Payload);

    private static Dictionary<string, object?> Unwrap(object? payload)
    {
        if (payload is IReadOnlyDictionary<string, object?> dict &&
            dict.GetValueOrDefault(MekikKey) is IReadOnlyDictionary<string, object?> meta)
        {
            var rest = dict.Where(kv => kv.Key != MekikKey).ToDictionary(kv => kv.Key, kv => kv.Value);
            var data = new Dictionary<string, object?> { ["payload"] = rest };
            if (meta.GetValueOrDefault("ui") is { } ui) data["ui"] = ui;
            if (meta.GetValueOrDefault("actions") is { } actions) data["actions"] = actions;
            return data;
        }
        return new Dictionary<string, object?> { ["payload"] = payload };
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private static readonly IReadOnlyDictionary<string, object?> EmptyState = new Dictionary<string, object?>();

    /// <summary>A token is ilmek's <see cref="TokenChunk"/> record OR a `{type:"token",text}` object.</summary>
    private static bool TryTokenText(object? payload, out string text)
    {
        switch (payload)
        {
            case TokenChunk tc:
                text = tc.Text;
                return true;
            case IReadOnlyDictionary<string, object?> d when d.GetValueOrDefault("type") is "token" && d.GetValueOrDefault("text") is string s:
                text = s;
                return true;
            default:
                text = "";
                return false;
        }
    }

    private static string FormatErrors(IReadOnlyList<(string Node, Exception Error)>? errors)
    {
        if (errors is null || errors.Count == 0) return "the run failed";
        return string.Join("; ", errors.Select(e => $"{e.Node}: {e.Error.Message}"));
    }
}

/// <summary>
/// Drive a whole recorded event list through a fresh mapper — the fixture entry
/// point. Production streams events one at a time through <see cref="TurnMapper.Map"/>.
/// </summary>
public static class Mapper
{
    public static List<Dictionary<string, object?>> EventToFrames(IEnumerable<IlmekEvent> events, TurnMapperDeps deps)
    {
        var mapper = new TurnMapper(deps);
        var frames = new List<Dictionary<string, object?>>();
        foreach (var ev in events) frames.AddRange(mapper.Map(ev));
        return frames;
    }
}
